docs/SPEC.md

```diff
@@ -839,6 +833,29 @@ Side effects can occur in two ways.
 
 A function produces side effects when it is explicitly declared to have side effects or when it calls a function that itself produces side effects.
 
+## Package initialization
+
+A package with no imports is initialized by first initializing and assigning all declared options and then subsequently initializing and assigning all declared variables.
+In particular all package options are initialized before any variables.
+Options cannot depend on any variables declared in the same package block.
+If a package has imports, all imported packages are initialized before any options or variables.
+
+### Variable initialization
+
+Variables are initialized and assigned the same value regardless of the order they are declared in the source.
+A variable cannot have a direct or indirect dependency on itself.
+
+### Option initialization
+
+An option cannot have dependencies on any other options assigned in the same package block.
+
+### Package imports
+
+Packages imported in the same file block are initialized in declaration order.
+A package cannot be imported twice in the same file block.
+A package will only be initialized once even if it is imported multiple times (across file blocks) in a package.
+Initializing imported packages must be deterministic.
+To be more precise, after all imported packages are initialized, each imported option must be assigned the same value regardless of the order in which package files are presented to an interpreter.
```

Comment:
I think we could simplify these three sections into a single ordered list followed by some additional notes an constraints. I am thinking something like this:

```
## Package initialization

Variables options and imports are all initialized in a specific order:

1. All imported packages are initialized following this package initialization process. 
2. All options in the current package block are assigned regardless of order. An option cannot have dependencies on any other options assigned in the same package block.
3. All variables in the current package block are assigned regardless of order. A variable cannot have a direct or indirect dependency on itself.


A package cannot be imported twice in the same file block. A package will only be initialized once across all files blocks in all packages. 

Initializing imported packages must be deterministic. 
Specifically after all imported packages are initialized, each option must be assigned the same value.
Packages imported in the same file block are initialized in declaration order. 
Packages imported across different file blocks have no known order. 
When a set of imports modify the same option, they must be ordered by placing them in the same file block. 

```

I have tried to capture all the constraints restrictions that we need. 


> A package cannot be imported twice in the same file block. 

Do we need this? Seems like it would just be a noop if it occurred.